tests/prototypes/prototype_mini.py

```diff
@@ -0,0 +1,508 @@
+            await self.commit_process()
+
+        self.info("Data streaming finished...")
+
+
+InputType = TypeVar("InputType")
+OutputType = TypeVar("OutputType")
+
+
+class DataReductionApp(BaseApp, Generic[InputType, OutputType]):
+    workflow: Workflow
+    input_pipe: List[InputType]
+    output_pipe: List[OutputType]
+
+    def __init__(self) -> None:
+        self.input_type = self._retrieve_type_arg('input_pipe')
+        self.output_type = self._retrieve_type_arg('output_pipe')
+        self.first_pulse_time: FirstPulseTime
+        super().__init__()
+
+    @classmethod
+    def _retrieve_type_arg(cls, attr_name: str) -> type:
+        """
+        Retrieve type arguments of an attribute with generic type.
+        It is only for retrieving input/output pipe type.
+
+        >>> class C(DataReductionApp):
+        ...   attr0: list[int]
+        ...
+        >>> C._retrieve_type_arg('attr0')
+        <class 'int'>
+        """
+        from typing import get_args, get_type_hints
+
+        if not (attr_type := get_type_hints(cls).get(attr_name)):
+            raise ValueError(
+                f"Class {cls} does not have an attribute "
+                f"{attr_name} or it is missing type annotation."
+            )
+        elif not (type_args := get_args(attr_type)):
+            raise TypeError(f"Attribute {attr_name} does not have any type arguments.")
+        else:
+            return type_args[0]
+
+    def format_received(self, data: InputType) -> str:
+        return str(data)
+
+    async def process_every_data(self, data: InputType) -> None:
+        self.debug("Received, %s", self.format_received(data))
+        with self.workflow.constant_provider(self.input_type, data):
+            self.output_pipe.append(self.workflow[self.output_type])
+
+        await self.commit_process()
+
+    def process_first_data(self, data: InputType) -> None:
+        ...
+
+    async def run(self) -> None:
+        data_monitor = self.data_pipe_monitor(self.input_pipe, target_size=1)
+        if not self.target_count_reached and await data_monitor():
+            data = self.input_pipe.pop(0)
+            self.process_first_data(data)
+            await self.process_every_data(data)
+
+        while not self.target_count_reached and await data_monitor():
+            data = self.input_pipe.pop(0)
+            await self.process_every_data(data)
+
+        self.info("No more data coming in. Finishing ...")
+
+
+class DataMerge(DataReductionApp[InputType, OutputType]):
+    input_pipe: List[Events]
+    output_pipe: List[MergedData]
+
+    def format_received(self, data: Any) -> str:
+        return f"{len(data)} pieces of {self.input_type.__name__}"
+
+    def process_first_data(self, data: Events) -> None:
+        sample_event = data[0]
+        first_pulse_time = sample_event.coords['event_time_zero'][0]
+        self.workflow.providers[FirstPulseTime] = lambda: first_pulse_time
```

Comment:
`DataReductionApp` has separate `process_first_data` method so that it can retrieve the `FirstPulseTime` (which is the `event_time_zero` of the first incoming data in this case).

And then it updates the `workflow` providers to return the retrieved value.